src/Interpreters/Cache/QueryConditionCache.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <Storages/MergeTree/MergeTreeData.h>
+
+namespace DB
+{
+
+using MarkFilter = std::vector<bool>;
+
+class IMergeTreeDataPart;
+
+/// Cache the mark filter corresponding to the query condition,
+/// which helps to quickly filter out useless Marks and speed up the query when the index is not hit.
+class QueryConditionCache
+{
+public:
+    QueryConditionCache(size_t max_count_)
+        : cache(std::numeric_limits<size_t>::max(), max_count_)
+    {}
+
+    /// Read the filter and return empty if it does not exist.
+    std::optional<MarkFilter> read(const std::shared_ptr<const IMergeTreeDataPart> & data_part, size_t condition_id);
+
+    /// Take out the mark filter corresponding to the query condition and set it to false on the corresponding mark.
+    void write(const MergeTreeDataPartPtr & data_part, size_t condition_id, const MarkRanges & mark_ranges);
+
+    void clear() { cache.clear(); }
+
+    void updateConfiguration(size_t max_entries) { cache.setMaxCount(max_entries); }
+
+private:
+    struct Key
+    {
+        const UUID table_id;
+        const String part_name;
+        const size_t condition_id;
+
+        bool operator==(const Key & other) const { return table_id == other.table_id && part_name == other.part_name && condition_id == other.condition_id; }
+    };
+
+    struct KeyHasher
+    {
+        size_t operator()(const Key & key) const;
+    };
+
+    struct Entry
+    {
+        /// If the mark at the corresponding position is false, there is no need to read.
+        MarkFilter mark_filter;
+        std::mutex mutex;
```

Comment:
Because a Part will be read by multiple threads at the same time, different bits of the `MarkFilter` will be updated concurrently. So a lock is added to each `MarkFilter`.